design_docs/2023_07_20_dynamic_shader_parameter_amount.md

```diff
@@ -0,0 +1,102 @@
+# Passing variables to the shader
+
+We need to be able to pass an arbitrary number of textures and parameters to the custom shaders. For parameters, we need to be able to pass some "global parameters" -- constants used in all videos, as well as some per-video data.
+
+## Possible solutions
+
+### Textures
+
+Unfortunately, it turns out that binding texture arrays is only possible when either
+
+- Textures have the same resolution
+
+or
+
+- The shader always accepts a constant number of textures
+
+This means we need to always bind a constant number of textures. If user's shader needs less textures than the length of the array, we can just fill the rest of it with references to a 1x1 texture which shouldn't harm performance in any way.
+
+Obviously we also need to pass the amount of textures that are actually present in the array. We should use a push constant for this. Push constants are a very efficient mechanism for sending a very small amount of parameters to the shader.
+
+Bringing it all together, the structure of variables declared in the user's shader could look like this:
+
+```wgsl
+@group(0) @binding(0) textures: binding_array<texture_2d<f32>, 16>;
+var<push_constant> textures_len: u32;
+```
+
+### Shader parameters
+
+Binding const-sized uniform/storage buffers works on every backend. This would work approximately like this in the shaders (ofc the len can be defined in a different way, e.g. as the length of the texture array):
+
+```wgsl
+struct MyList {
+    // you declare an array with a constant, large size
+    per_video_data: array<MyStruct, 64>
+    // and then in some way specify how much of this buffer is filled
+    per_video_data_len: u32
```

Comment:
1. From the comment, I get the impression, that this represents how many structs are in the array, while this variable name suggests something like a single `MyStruct` size/len. Maybe something like `videos_count` would be better?
2. Wouldn't wgpu memory layout force us to fill up this array with exactly `sizeof(MyStruct) * 64` bytes? Have you tried passing it as uniform without filling this list up entirely?

Reply:
I did try and it works, but the array should be the last element in the struct. Then we can only send, for example, the first half of the buffer.

Reply:
Also this is an example struct, not the API proposal for all users to use, each shader will declare it's own structs.